hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java

```diff
@@ -2358,7 +2366,7 @@ private KeyOutputStream.Builder createKeyOutputStream(
         .enableUnsafeByteBufferConversion(unsafeByteBufferConversion)
         .setConfig(clientConfig)
         .setAtomicKeyCreation(isS3GRequest.get())
-        .setClientMetrics(clientMetrics)
+        .setBlockOutPutStreamResourceProvider(blockOutPutStreamResourceProvider)
```

Comment:
Maybe I'm missing something, but it seems the same `ecWriteExecutor` is used for both EC and Ratis keys.

Reply:
Hi @adoroszlai @xichen01 , Thanks work on this, I think the reason for the current SingleThreadExecutor is because it uses asynchrony in an OutputStream and needs to guarantee the order of execution. If that's the case, is switching to a multi-threaded Executor still a guarantee?

Reply:
> Maybe I'm missing something, but it seems the same `ecWriteExecutor` is used for both EC and Ratis keys:
> 
> https://github.com/apache/ozone/blob/f695a374420a49728e2be352534cbdbe48ece0d9/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L319-L320
> 
> @xichen01 Shouldn't we use a different resource provider for Ratis keys?

@adoroszlai Yes, since Ratis Key and EC Key reuse class `BlockOutputStream`, the EC Key and Ratis Key both use this `ecWriteExecutor`. I think that's acceptable, since `ecWriteExecutor` just provides thread resources, not logic. And the Ratis key also benefits from not having to create a new thread for each Block. But mybe we need rename the `ecWriteExecutor` to a more normal name.

Reply:
Thanks for the clarification.

Reply:
> Hi @adoroszlai @xichen01 , Thanks work on this, I think the reason for the current SingleThreadExecutor is because it uses asynchrony in an OutputStream and needs to guarantee the order of execution. If that's the case, is switching to a multi-threaded Executor still a guarantee?

@guohao-rosicky. I think it's okay here, because `ecWriteExecutor` is just used to wait for an RPC response, not to send a request. the thread that sends request was always a fixed single thread.
@adoroszlai how do you think?

Reply:
@xichen01 This part of response processing may not be fully safe for multi-thread use:

https://github.com/apache/ozone/blob/6f507c9bb1a66e667006f8609a3e900950fedd20/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockOutputStream.java#L490-L493

https://github.com/apache/ozone/blob/6f507c9bb1a66e667006f8609a3e900950fedd20/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/RatisBlockOutputStream.java#L108-L109

https://github.com/apache/ozone/blob/6f507c9bb1a66e667006f8609a3e900950fedd20/hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/AbstractCommitWatcher.java#L76-L79

due to the `LinkedList`

Reply:
Add a `synchronized` for `thenApplyAsync`.

Reply:
Thanks a lot @xichen01 for updating the patch.  Most of the updates look good, but I think this part needs to improved.  I think it would be better to replace the `LinkedList` in `AbstractCommitWatcher` with a `CopyOnWriteArrayList`, rather than synchronizing `thenApplyAsync` completely.

CC @szetszwo

Reply:
Could the `synchronized` be moved to `AbstractCommitWatcher`?  It is a much smaller class.

Reply:
I noticed that the closure accesses a lot of external variables, which could be a potential thread-safety issue, so I just added a "big" `synchronized` here, which ensures that even if someone changes the code later on, there's probably no thread-safety issue, and the `thenApplyAsync` just sets/creates the variable, which is supposed to be lightweight.
So, I think the current approach is acceptable, what do you think?

Reply:
> closure accesses a lot of external variables, which could be a potential thread-safety issue

I have the following concerns:

1. Are the same external variables protected by the same object elsewhere?  If not, there is no point in making this synchronized.
2. `validateResponse` only accesses an atomic reference and the response object itself.  Neither of those need to be synced.
3. Protobuf conversion (`BlockID.getFromProtobuf`) should not be performed while holding the lock.
4. `(this.)blockID` is also atomic, does not need the lock.
5. logging should not be performed while holding the lock.

So in the end I think only `updateCommitInfo` needs to be protected.

Reply:
`synchronized` has been moved to `AbstractCommitWatcher`.